src/smoke_test.rs

```diff
@@ -0,0 +1,113 @@
+}
+
+impl SmokeTester {
+    pub(crate) fn new(paths: &[PathBuf]) -> Result<Self, Error> {
+        let addr = SocketAddr::from(([127, 0, 0, 1], 0));
+
+        let paths = Arc::new(paths.to_vec());
+        let service = hyper::service::make_service_fn(move |_| {
+            let paths = paths.clone();
+            async move {
+                Ok::<_, Error>(hyper::service::service_fn(move |req| {
+                    server_handler(req, paths.clone())
+                }))
+            }
+        });
+
+        let (shutdown_send, shutdown_recv) = tokio::sync::oneshot::channel::<()>();
+
+        let runtime = Runtime::new()?;
+        let (server, server_addr) = runtime.enter(|| {
+            let server = Server::bind(&addr).serve(service);
+            let server_addr = server.local_addr();
+            let server = server.with_graceful_shutdown(async {
+                shutdown_recv.await.ok();
+            });
+            (server, server_addr)
+        });
+        runtime.spawn(server);
+
+        Ok(Self {
+            runtime,
+            server_addr,
+            shutdown_send,
+        })
+    }
+
+    pub(crate) fn server_addr(&self) -> SocketAddr {
+        self.server_addr
+    }
+
+    pub(crate) fn test(self, channel: &Channel) -> Result<(), Error> {
+        let tempdir = TempDir::new()?;
+        let cargo_dir = tempdir.path().join("sample-crate");
+        std::fs::create_dir_all(&cargo_dir)?;
+
+        let cargo = |args: &[&str]| {
+            crate::run(
+                Command::new("cargo")
+                    .arg(format!("+{}", channel))
+                    .args(args)
+                    .env("USER", "root")
+                    .current_dir(&cargo_dir),
+            )
+        };
+        let rustup = |args: &[&str]| {
+            crate::run(
+                Command::new("rustup")
+                    .env("RUSTUP_DIST_SERVER", format!("http://{}", self.server_addr))
+                    .args(args),
+            )
+        };
+
+        rustup(&["toolchain", "remove", &channel.to_string()])?;
+        rustup(&["toolchain", "install", &channel.to_string()])?;
+        cargo(&["init", "--bin", "."])?;
+        cargo(&["run"])?;
+
+        // Finally shut down the HTTP server and the tokio reactor.
+        self.shutdown_send
+            .send(())
+            .expect("failed to send shutdown message to the server");
+        self.runtime.shutdown_background();
+
+        Ok(())
+    }
+}
+
+async fn server_handler(
+    req: Request<Body>,
+    paths: Arc<Vec<PathBuf>>,
+) -> Result<Response<Body>, Error> {
+    let file_name = match req.uri().path().split('/').last() {
+        Some(file_name) => file_name,
+        None => return not_found(),
+    };
+    for directory in &*paths {
+        let path = directory.join(file_name);
+        if path.is_file() {
+            let content = tokio::fs::read(&path).await?;
+            return Ok(Response::new(content.into()));
+        }
+    }
+    not_found()
+}
```

Comment:
I am wondering if we really need to add the hyper (relatively heavy) dependency for this. Could we rely on (e.g.) `python -m http.server` for this? We're just serving static files AFAICT and it really feels quite annoying to add what is likely multiple minutes to CI / minute or two for local development of building hyper just for that.

Reply:
I don't think this will be the only use of hyper. Reqwest replacing curl and rusoto replacing a lot of the shelling out to awscli would also be possible if we start depending on hyper.

Reply:
Oh, also, we can't really use `python -m http.server`, as we're serving files from either the dl dir or the manifest dir.

Reply:
Ok, seems unfortunate but ultimately reasonable. I wish we had a more minimal (threadless, sync, etc.) hyper for this sort of use case.